modal/stub.py

```diff
@@ -382,11 +382,11 @@ async def serve(self, client=None, stdout=None, show_progress=None, timeout=None
                     output_mgr.print_if_visible(f"⚡️ Updating app {existing_app_id}...")
 
                 async with self._run(client, output_mgr, existing_app_id, mode=StubRunMode.SERVE) as app:
+                    client.set_pre_stop(app.disconnect)
                     existing_app_id = app.app_id
                     event = await event_agen.__anext__()
         finally:
             await event_agen.aclose()
-            await app.disconnect()
```

Comment:
wouldn't the client disconnect happen after the async destructor is called? is the problem that coroutines registered using `on_shutdown` may run before async exit handlers?

Reply:
is it basically just an async race? maybe a slightly cleaner thing would be to "borrow" the client using a context manager that prevents it from being closed until all borrows have been relinquished

Reply:
not super important, just some random ideas

Reply:
It seems not to run after. The status quo is the the async loop is shutdown before the disconnecting RPC can complete. 

I think the `on_shutdown` mechanism is receiving the cancellation and immediately scheduling the `coro` task in the runtime, but the `finally` destructor here doing the `app.disconnect()` is the last layer of an onion that begins with the `event_agen` coro getting the cancellation exception but has intermediate layers.

Pretty unpleasant.

Reply:
annoying yeah. I have some ideas how we could fix it, but it doesn't seem like very high ROI. Let's ship this PR.